pipeline/metadata/flatten.py

```diff
@@ -366,13 +370,56 @@ def _process_satellite_v2(self, scan: Any,
         'error': scan.get('error', None),
         'anomaly': scan['anomaly'],
         'success': not scan['connect_error'],
+        'controls_failed': not scan['passed_control'],
         'received': None,
         'rcode': [],
         'measurement_id': random_measurement_id
     }
     received_ips = scan.get('response')
     yield from self._process_received_ips(row, received_ips)
 
+  def _process_satellite_v2_responses(
+      self, scan: Any, random_measurement_id: str) -> Iterator[Row]:
+    """Process a line of Satellite response data.
+
+      Args:
+        scan: a loaded json object containing the parsed content of the line
+        random_measurement_id: a hex id identifying this individual measurement
+
+      Yields:
+        Rows
+    """
+    row = {
+        'domain': scan['test_url'],
+        'category': self.category_matcher.match_url(scan['test_url']),
+        'ip': scan['vp'],
+        'anomaly': None,
+        'success': not scan['connect_error'],
+        'controls_failed': not scan['passed_control'],
+        'measurement_id': random_measurement_id
+    }
+    responses = scan.get('response', [])
+    if responses:
+      row['date'] = responses[0]['start_time'][:10]
+      row['start_time'] = format_timestamp(responses[0]['start_time'])
+      row['end_time'] = format_timestamp(responses[-1]['end_time'])
+      row['rcode'] = [str(response['rcode']) for response in responses]
+      errors = [
+          response['error']
+          for response in responses
+          if response['error'] and response['error'] != 'null'
+      ]
+      row['error'] = ' | '.join(errors) if errors else None
+      for response in responses:
+        if response['url'] == row['domain']:
+          # Check response for test domain
+          if response['rcode'] == 0 and response['has_type_a']:
+            # Valid answers
+            row['has_type_a'] = True
+            for ip in response['response']:
+              row['received'] = {'ip': ip}
+              yield row
```

Comment:
I'm a bit worried that this yield is emitting the same row object multiple times while just overwriting its ip field. You may need to make a copy of the row before emitting here, see the comment below on the test.

Reply:
This was the issue! `_process_received_ips` also had the same problem. I've updated these to yield a copy of the row.